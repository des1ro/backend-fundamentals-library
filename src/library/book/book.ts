export class Book {
  constructor(
    private readonly title: string,
    private readonly author: string,
    private readonly year: number,
    private readonly id: number
  ) {}
  getId(): number {
    return this.id;
  }
  getTitle(): string {
    return this.title;
  }
  getAuthor(): string {
    return this.author;
  }
  getYear(): number {
    return this.year;
  }
}
