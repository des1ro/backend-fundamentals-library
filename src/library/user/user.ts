import { randomUUID } from "crypto";

export class User {
  private readonly uuid: string = randomUUID();
  private credit: number = 0;
  constructor(private readonly name: string) {}
  getName(): string {
    return this.name;
  }
  getUuid(): string {
    return this.uuid;
  }
  resetCredit() {
    this.credit = 0; // Czy tu powinienem wrzucić credit do construktora, (private readonly credit) i chcąc zmiany tworzyć new User?
  }
  getCredit() {
    return this.credit;
  }
  setCredit(credit: number): void {
    this.credit = credit;
  }
}
