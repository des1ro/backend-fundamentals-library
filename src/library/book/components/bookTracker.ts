import { LibraryError } from "../../exceptions/library.exceptions";
import { User } from "../../user/user";

export class BookTracker {
  private borrowedBy: User | null = null;
  borrow(user: User): void {
    if (this.borrowedBy) {
      throw new LibraryError({
        name: "BOOKTRACKER_ERROR",
        message: "Book already borrowed",
      });
    }
    this.borrowedBy = user;
  }
  reset(): void {
    this.borrowedBy = null;
  }
  getUser(): User {
    if (!this.borrowedBy) {
      throw new LibraryError({
        name: "BOOKTRACKER_ERROR",
        message: "Book is not borrowed",
      });
    }
    return this.borrowedBy;
  }
}
