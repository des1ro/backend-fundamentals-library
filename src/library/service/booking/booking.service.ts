import { isAfter, addMonths } from "date-fns";
import { Book } from "../../../book/book.dto";
import { LibraryError } from "../../../exceptions/library.exceptions";
import { User } from "../../../user/user.dto";

export class Booking {
  private readonly penaltyPointsPerBook: number;
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {
    this.penaltyPointsPerBook = 10;
  }
  addBookToUser(book: Book): void {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    const bookBorrowDateLimit = addMonths(date, 1);
    const bookIsAfterDueDate = isAfter(new Date(), bookBorrowDateLimit);
    this.booksBorrowDate.delete(returnedBook);
    if (bookIsAfterDueDate) {
      const penaltyPoints =
        this.user.getPenaltyPoints() + this.penaltyPointsPerBook;
      this.user.setPenaltyPoints(penaltyPoints);
    }
  }
  getBooksDatesArray(): Date[] {
    const datesArray = Array.from(this.booksBorrowDate.values());
    return datesArray;
  }
}
