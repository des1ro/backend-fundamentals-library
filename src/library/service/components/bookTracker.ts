import { LibraryError } from "../../../exceptions/library.exceptions";
import { User } from "../../../user/user";
import { Book } from "../../../book/book";
export class BookTracker {
  private borrowedBy: User | null = null;
  constructor(readonly book: Book) {}
  borrow(user: User): void {
    if (this.borrowedBy) {
      throw new LibraryError({
        name: "BOOKTRACKER_ERROR",
        message: "Book already borrowed",
      });
    }
    this.borrowedBy = user;
  }
  reset(): void {
    this.borrowedBy = null;
  }
  getUser(): User {
    if (!this.borrowedBy) {
      throw new LibraryError({
        name: "BOOKTRACKER_ERROR",
        message: "Book is not borrowed",
      });
    }
    return this.borrowedBy;
  }
}
