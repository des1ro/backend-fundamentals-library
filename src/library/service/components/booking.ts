import { add, addMonths, differenceInMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
// sprawdzać przy oddawaniu czy książka po terminie
// wszystko w bookingu
// user ma mieć
export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {}
  getUser(): User {
    return this.user;
  }
  isUserBanned(
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): boolean {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      return true;
    }
    const userPenaltyPoints = this.checkUserPenaltyPoints(
      bookBorrowMonthsLimit,
      penaltyPointsLimit
    );
    if (userPenaltyPoints > 0) {
      const banMonths = userPenaltyPoints / 10;
      this.user.setDueDate(addMonths(new Date(), banMonths));
      return true;
    }
    return false;
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned(bookBorrowMonthsLimit, penaltyPointsLimit)) {
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERRMISION_DENIED",
      message: "User is banned!",
    });
  }
  checkUserPenaltyPoints(
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): number {
    let penaltyPoints = 0;
    this.booksDueDate.forEach((bookDate) => {
      const difference = differenceInMonths(new Date(), bookDate);
      if (difference >= bookBorrowMonthsLimit)
        penaltyPoints += penaltyPointsPerBook;
    });
    this.user.addPenaltyPoints(penaltyPoints);
    return this.user.getPenaltyPoints();
  }

  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): void {
    const isUserBanned = this.isUserBanned(
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (this.booksDueDate.has(returnedBook)) {
      const date = this.booksDueDate.get(returnedBook);
      if (date && isAfter(new Date(), addMonths(date, 1)) && isUserBanned) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
      }
      this.booksDueDate.delete(returnedBook);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  getBooksDueDate(): Map<Book, Date> {
    return this.booksDueDate;
  }
  getUserDueDate() {
    return this.user.getDueDate();
  }
  getUserPenaltyPoints() {
    return this.user.getPenaltyPoints();
  }
}
