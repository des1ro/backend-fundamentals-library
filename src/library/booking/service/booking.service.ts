import { isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
import { PenaltyPoints } from "./components/PenaltyPoints";
export class Booking {
  private readonly penaltyPoints: PenaltyPoints;
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {
    this.penaltyPoints = new PenaltyPoints(this.user);
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned()) {
      const penaltyPoints = this.penaltyPoints.calculatePenaltyPoints(
        this.booksDueDate,
        bookBorrowMonthsLimit,
        penaltyPointsPerBook
      );
      if (penaltyPoints > penaltyPointsLimit) {
        this.penaltyPoints.updateUserPenaltyPointsAndBan(penaltyPoints);
        throw new LibraryError({
          name: "PERMISION_DENIED",
          message: "User is banned!",
        });
      }
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERMISION_DENIED",
      message: "User is banned!",
    });
  }
  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number,
    penaltyPointsLimit: number
  ): void {
    const penaltyPoints = this.penaltyPoints.calculatePenaltyPoints(
      this.booksDueDate,
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (penaltyPoints > penaltyPointsLimit) {
      this.penaltyPoints.updateUserPenaltyPointsAndBan(penaltyPoints);
    }
    if (this.booksDueDate.has(returnedBook)) {
      this.booksDueDate.delete(returnedBook);
      if (this.isUserBanned() && penaltyPoints > 0) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
        return;
      }
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  has(book: Book): boolean {
    return this.booksDueDate.has(book);
  }
  private isUserBanned(): boolean {
    return isAfter(this.user.getDueDate(), new Date());
  }
}
