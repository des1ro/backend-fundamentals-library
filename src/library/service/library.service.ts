import { Book } from "../book/book";
import { LibraryError } from "../exceptions/library.exceptions";
import { User } from "../user/user";
import { Booking } from "./components/booking";
export class Library {
  private readonly bookBorrowMonthsLimit: number;
  private readonly penaltyPointsPerBook: number;
  constructor(
    private readonly libraryBooks = new Map<Book, boolean>(),
    private readonly bookings = new Map<User, Booking>()
  ) {
    this.bookBorrowMonthsLimit = 1;
    this.penaltyPointsPerBook = 10;
  }
  addBook(book: Book): void {
    this.libraryBooks.set(book, true);
  }
  deleteBook(book: Book): void {
    if (!this.libraryBooks.delete(book)) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not available in the library",
      });
    }
  }
  borrowBook(user: User, book: Book) {
    let booking = this.bookings.get(user);
    if (!booking) {
      booking = new Booking(user);
      this.bookings.set(user, booking);
    }
    booking.addBookToUser(
      book,
      this.bookBorrowMonthsLimit,
      this.penaltyPointsPerBook
    );
    this.libraryBooks.set(book, false);
  }
  returnBook(book: Book) {
    const booking = Array.from(this.bookings.values()).find((booking) =>
      booking.getBooksDueDate().has(book)
    );
    if (booking) {
      booking.returnBook(
        book,
        this.bookBorrowMonthsLimit,
        this.penaltyPointsPerBook
      );
      this.libraryBooks.set(book, true);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "Book not found",
    });
  }
  getBooks() {
    return this.libraryBooks;
  }
  getBookings() {
    return this.bookings;
  }
}
