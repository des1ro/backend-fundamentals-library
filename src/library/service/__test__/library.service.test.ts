import { Book } from "../../book/book";
import { LibraryError } from "../../exceptions/library.exceptions";
import { User } from "../../user/user";
import { Booking } from "../components/booking";
import { Library } from "../library.service";
jest.mock("crypto", () => ({
  randomUUID: jest.fn().mockReturnValue("test-uuid"),
}));
describe("Library test suite", () => {
  let objectUnderTest: Library;
  let mockedUserOne: User;
  let mockedUserTwo: User;
  let mockedBookOne: Book;
  let mockedBookTwo: Book;
  let mockedBookingOne: Booking;
  let mockedBookingTwo: Booking;
  let testDate: Date;
  let spy: jest.SpyInstance;
  const testBookBorrowDaysLimit = 30;
  beforeAll(() => {
    mockedBookOne = new Book("Test title One", "Test author One", 1992, 1);
    mockedBookTwo = new Book("Test title Two", "Test author Two", 1995, 2);
  });
  beforeEach(() => {
    mockedUserOne = new User("Test user one");
    mockedUserTwo = new User("Test user one");
    objectUnderTest = new Library(testBookBorrowDaysLimit);
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });
  it("Should add book properly", () => {
    //When
    const resultOne = objectUnderTest.getBooks();
    objectUnderTest.addBook(mockedBookOne);
    const resultTwo = objectUnderTest.getBooks();
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(resultTwo).toMatchSnapshot();
  });
  describe("Delete book test suite", () => {
    it("Should delete book properly", () => {
      //Given
      spy = jest.spyOn(objectUnderTest, "deleteBook");
      //When
      objectUnderTest.addBook(mockedBookOne);
      const resultOne = objectUnderTest.getBooks();
      objectUnderTest.deleteBook(mockedBookOne);
      const resultTwo = objectUnderTest.getBooks();
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(spy).toBeCalledWith(mockedBookOne);
    });
    it("Should throw an LibraryError if book isn't in library", () => {
      //When
      const result = objectUnderTest.getBooks();
      //Then
      expect(result).toMatchSnapshot();
      expect(() => objectUnderTest.deleteBook(mockedBookOne)).toThrow(
        LibraryError
      );
    });
  });
  describe("Update booking test suite", () => {
    it("Should only add credit if user credit < 10", () => {
      //Given
      jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
      objectUnderTest = new Library(testBookBorrowDaysLimit);
      mockedBookingOne = new Booking(mockedUserOne);
      //When
      const spy = jest
        .spyOn(mockedBookingOne, "getUserCredit")
        .mockReturnValue(9);
      objectUnderTest.addBooking(mockedBookingOne);
      objectUnderTest.updateBookings();
      const result = objectUnderTest.getBookings();
      //Then
      expect(spy).toBeCalledTimes(1);
      expect(result).toMatchSnapshot();
    });
    it("Should add user to block list if credit >= 10 ", () => {
      //Given
      jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
      objectUnderTest = new Library(testBookBorrowDaysLimit);
      mockedBookingOne = new Booking(mockedUserOne);
      //When
      jest.spyOn(mockedBookingOne, "getUserCredit").mockReturnValue(10);
      objectUnderTest.addBooking(mockedBookingOne);
      const resultOne = objectUnderTest.getBookings();
      objectUnderTest.updateBookings();
      const resultTwo = objectUnderTest.getBookings();
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
    });
  });

  it("Should return booking if it's in bookings", () => {
    //Given
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    mockedBookingOne = new Booking(mockedUserOne);
    mockedBookingTwo = new Booking(mockedUserTwo);
    objectUnderTest.addBooking(mockedBookingOne);
    objectUnderTest.addBooking(mockedBookingTwo);
    //When
    const resultOne = objectUnderTest.getBooking(mockedUserOne);
    const resultTwo = objectUnderTest.getBookings();
    //Then
    expect(resultOne).toBe(mockedBookingOne);
    expect(resultTwo).toMatchSnapshot();
  });

  it("Should return new booking if isn't in bookings ", () => {
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    mockedBookingOne = new Booking(mockedUserOne);
    mockedBookingTwo = new Booking(mockedUserTwo);
    objectUnderTest.addBooking(mockedBookingTwo);
    //When
    const resultOne = objectUnderTest.getBooking(mockedUserOne);
    const resultTwo = objectUnderTest.getBookings();
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(resultTwo).toMatchSnapshot();
  });
  it("Should delete book form library and add to booking if user isn't banned ", () => {
    //Given
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    //When
    objectUnderTest.addBook(mockedBookOne);
    objectUnderTest.borrowBook(mockedUserOne, mockedBookOne);
    const resultOne = objectUnderTest.getBooks();
    const resultTwo = objectUnderTest.getBookings();
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(resultTwo).toMatchSnapshot();
  });
  it("Should throw an LibraryError if banned user want to borrow book", () => {
    //Given
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    objectUnderTest = new Library(testBookBorrowDaysLimit);
    testDate = new Date(2023, 8, 20, 12, 30, 0, 0);
    mockedBookingOne = new Booking(mockedUserOne, testDate);
    //When
    objectUnderTest.addBooking(mockedBookingOne);
    objectUnderTest.addBook(mockedBookOne);
    const resultOne = objectUnderTest.getBooking(mockedUserOne);
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(() =>
      objectUnderTest.borrowBook(mockedUserOne, mockedBookOne)
    ).toThrow(LibraryError);
  });
  it("Should return book properly", () => {
    //Given
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    testDate = new Date();
    const books = new Map<Book, Date>();
    books.set(mockedBookOne, testDate);
    const testBooking = new Booking(mockedUserOne, testDate, books);
    //When
    objectUnderTest.addBooking(testBooking);
    const resultOne = objectUnderTest.getBooking(mockedUserOne);
    //Then
    expect(resultOne).toMatchSnapshot();
  });
  it("Should throw an LibraryError on returnig book, which isn't in bookings", () => {
    //Given
    jest.setSystemTime(new Date(2023, 6, 20, 12, 30, 0, 0));
    testDate = new Date();
    const books = new Map<Book, Date>();
    books.set(mockedBookOne, testDate);
    const testBooking = new Booking(mockedUserOne, testDate, books);
    //When
    objectUnderTest.addBooking(testBooking);
    const resultOne = objectUnderTest.getBooking(mockedUserOne);
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(() => objectUnderTest.returnBook(mockedBookTwo)).toThrow(
      LibraryError
    );
  });
});
