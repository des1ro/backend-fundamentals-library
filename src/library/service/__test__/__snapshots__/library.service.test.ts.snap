// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library test suite Borrow book suite test Should borrow book and set available to false 1`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => false,
}
`;

exports[`Library test suite Borrow book suite test Should borrow book and set available to false 2`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => false,
}
`;

exports[`Library test suite Borrow book suite test Should create new booking if isn't in bookings  1`] = `
Map {
  User {
    "dueDate": 2023-01-20T11:30:00.000Z,
    "name": "Test user one",
    "penaltyPoints": 0,
    "uuid": "test-uuid",
  } => Booking {
    "booksDueDate": Map {
      Book {
        "author": "Test author One",
        "id": 1,
        "title": "Test title One",
        "year": 1992,
      } => 2023-01-20T11:30:00.000Z,
    },
    "user": User {
      "dueDate": 2023-01-20T11:30:00.000Z,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    },
  },
}
`;

exports[`Library test suite Borrow book suite test Should create new booking if isn't in bookings  2`] = `
Map {
  User {
    "dueDate": 2023-01-20T11:30:00.000Z,
    "name": "Test user one",
    "penaltyPoints": 0,
    "uuid": "test-uuid",
  } => Booking {
    "booksDueDate": Map {
      Book {
        "author": "Test author One",
        "id": 1,
        "title": "Test title One",
        "year": 1992,
      } => 2023-01-20T11:30:00.000Z,
    },
    "user": User {
      "dueDate": 2023-01-20T11:30:00.000Z,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should delete book properly 1`] = `Map {}`;

exports[`Library test suite Delete book test suite Should delete book properly 2`] = `Map {}`;

exports[`Library test suite Delete book test suite Should throw an LibraryError if book isn't in library 1`] = `Map {}`;

exports[`Library test suite Return book suite test Should return book properly and set to be available in library 1`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => true,
}
`;

exports[`Library test suite Return book suite test Should throw an LibraryError on return wrong book 1`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => false,
}
`;

exports[`Library test suite Return book suite test Should throw an LibraryError on return wrong book 2`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => false,
}
`;

exports[`Library test suite Should add book properly 1`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => true,
}
`;

exports[`Library test suite Should add book properly 2`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => true,
}
`;
